File: parser/src/main/java/com/github/fburato/highwheelmodules/bytecodeparser/ClassPathParser.java
package com.github.fburato.highwheelmodules.bytecodeparser;


import java.io.IOException;
import java.io.InputStream;
import java.util.function.Predicate;

import org.objectweb.asm.ClassReader;
import com.github.fburato.highwheelmodules.model.classpath.AccessVisitor;
import com.github.fburato.highwheelmodules.model.classpath.ClassParser;
import com.github.fburato.highwheelmodules.model.classpath.ClasspathRoot;
import com.github.fburato.highwheelmodules.model.bytecode.ElementName;

public class ClassPathParser implements ClassParser {

  private final Predicate<ElementName> filter;
  private final NameTransformer nameTransformer;
  
  public ClassPathParser(final Predicate<ElementName> filter) {
    this(filter, new CollapseInnerClassesNameTransformer());
  }

  public ClassPathParser(final Predicate<ElementName> filter, final NameTransformer nameTransformer) {
    this.filter = filter;
    this.nameTransformer = nameTransformer;
  }

  public void parse(final ClasspathRoot classes, final AccessVisitor v) throws IOException {

    for (final ElementName each : classes.classNames()) {
      if (this.filter.test(each)) {
        parseClass(classes, v, each);
      }
    }

  }

  private void parseClass(final ClasspathRoot cp, final AccessVisitor dv,
      final ElementName each) throws IOException {
    final InputStream is = cp.getData(each);
    try {
      final ClassReader reader = new ClassReader(is);
      final DependencyClassVisitor cv = new DependencyClassVisitor(null,
          new FilteringDecorator(dv, this.filter), nameTransformer);
      reader.accept(cv, 0);
    } finally {
      is.close();
    }

  }

}

File: parser/src/main/scala/com/github/fburato/highwheelmodules/bytecodeparser/ClassPathParser.scala
package com.github.fburato.highwheelmodules.bytecodeparser

import com.github.fburato.highwheelmodules.model.bytecode.ElementName
import com.github.fburato.highwheelmodules.model.classpath._
import com.github.fburato.highwheelmodules.utils.TryUtils._
import org.objectweb.asm.ClassReader

import java.io.InputStream
import scala.util.{Success, Try, Using}

class ClassPathParser(filter: ElementName => Boolean) extends ClassParser {

  private val nameTransformer: NameTransformer = s =>
    if (s.contains("$")) {
      ElementName.fromString(s.substring(0, s.indexOf('$')))
    } else {
      ElementName.fromString(s)
    }

  override def parse(cpr: ClasspathRoot, accessVisitor: AccessVisitor): Try[Unit] = {
    val dependencyClassVisitor = new DependencyClassVisitor(
      null,
      new FilteringDecorator(accessVisitor, filter),
      nameTransformer
    )

    def readAndVisitClassStream(is: InputStream): Try[Unit] =
      for {
        classReader <- Try(new ClassReader(is))
        _ <- Try(classReader.accept(dependencyClassVisitor, 0))
      } yield ()

    def parseElement(elementName: ElementName): Try[Unit] =
      for {
        maybeStream <- cpr.getData(elementName)
        compute <- maybeStream.map(stream => Using(stream)(readAndVisitClassStream)).getOrElse(Success(Success()))
        _ <- compute
      } yield ()

    for {
      classes <- cpr.classNames
      _ <- sequence(classes.filter(filter).map(parseElement))
    } yield ()
  }
}
